### ISSUE TITLE
New tags attribute causes error when unpickling older models

### ISSUE THREAD

**Issue by bgyori** [pysb/pysb#467]
The PR https://github.com/pysb/pysb/pull/464/ adds a new `tags` attribute to the Model class. For any model that was pickled before this PR was merged, we get an error in the setstate code:

```python
    def __setstate__(self, state):
        # restore the 'model' weakrefs on all components
        self.__dict__.update(state)
        for c in self.all_components():
            c.model = weakref.ref(self)
```
like 
```
  File "/home/travis/virtualenv/python3.6.7/lib/python3.6/site-packages/pysb/core.py", line 1845, in __setstate__
    for c in self.all_components():
  File "/home/travis/virtualenv/python3.6.7/lib/python3.6/site-packages/pysb/core.py", line 1914, in all_components
    for cset in self.all_component_sets():
  File "/home/travis/virtualenv/python3.6.7/lib/python3.6/site-packages/pysb/core.py", line 1908, in all_component_sets
    sets = [getattr(self, name) for name in set_names]
  File "/home/travis/virtualenv/python3.6.7/lib/python3.6/site-packages/pysb/core.py", line 1908, in <listcomp>
    sets = [getattr(self, name) for name in set_names]
AttributeError: 'Model' object has no attribute 'tags'
```
Should anything in PySB be done to mitigate this, or is this purely a downstream concern?

**Comment by alubbock** [comment]
This should get fixed in PySB. Apologies for the oversight - fortunately it's a one-line fix. See #468.

**Comment by jmuhlich** [comment]
There are going to be lots of other problems using a Model unpickled from an older PySB version. For example @bgyori's unpickled models will crash the simulators since _derived_parameters and _derived_expressions won't exist. The problem isn't even limited to this PR or even to the Model class -- any class that changes its structure will be a problem. See here:

https://wiki.python.org/moin/UsingPickle#Discussion

> Pickles can cause problems if you save a pickle, then update your code and read the pickle in. Attribute added to your \_\_init\_\_ may not be present in the unpickled object ... For this reason, you should be wary of using pickles for long-term storage where the underlying code is not highly stable.

I think handling this sort of thing does belong in the end-user's code. Maybe you could write a pickle upgrade script that monkey-patches Model.__setstate__ to add all the new attributes?

**Comment by bgyori** [comment]
Okay, I agree in theory, but there really aren't all that many options other than pickles that we can use to store Models. The only format I'm aware of to store a PySB Model other than a pickle is the PySB-flat format which creates a standalone .py file to represent a Model. That of course is not ideal for the automated creation/storage/reuse of models since it requires importing or exec-ing code. This suggests that PySB might benefit from a Model serialization approach like JSON - without that, we are stuck with pickles.

**Comment by alubbock** [comment]
I've thought about a JSON serializer for PySB models before, e.g. for use in the HDF5 `SimulationResult.save()` which stores a pickled model. Although it's annoying to introduce a new model "format", a JSON serializer seems like a reasonable option for embedding/storing models, or transmitting them (e.g. as part of HTTP APIs).

**Comment by bgyori** [comment]
Yes, the transmission of models is another issue that we've run into, and had to work around since without a better alternative, one would only be able to serialize into PySB-flat, and then exec the code from the transmitted string - obviously not a good idea.

**Comment by alubbock** [comment]
Fixed by #469 in PySB v1.11.0 for `SimulationResult`. For standalone use, it's now recommended to use the new JSON exporter and importer. For legacy pickled models, a shim can be used to load models saved in PySB <1.10 in PySB >=1.10:

```python
from pysb.simulator.base import _patch_model_setstate

with open('model.pkl', 'rb') as f:
    with _patch_model_setstate():
        pickle.load(f)
```
